File: backend/routers/vms.py
"""VMS Management routes"""
from fastapi import APIRouter, HTTPException, Depends, Header
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel

from database import db
from models import VMSServer, VMSServerCreate, VMSServerUpdate, Camera, ImportCamerasRequest
from vms_utils import fetch_vms_data, parse_counter_xml, parse_queue_xml
from auth import require_auth, resolve_tenant_filter, resolve_tenant_id

router = APIRouter(prefix="/vms", tags=["VMS"])


class GroupRenameRequest(BaseModel):
    new_name: str


@router.post("")
async def create_vms(input: VMSServerCreate, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    data = input.model_dump()
    data["tenant_id"] = resolve_tenant_id(user, x_tenant_id)
    vms_obj = VMSServer(**data)
    doc = vms_obj.model_dump()
    doc["created_at"] = doc["created_at"].isoformat()
    await db.vms_servers.insert_one(doc)
    doc.pop("_id", None)
    return doc


@router.get("")
async def get_vms_list(user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    servers = await db.vms_servers.find(resolve_tenant_filter(user, x_tenant_id), {"_id": 0}).to_list(100)
    for s in servers:
        if isinstance(s.get("created_at"), str):
            s["created_at"] = s["created_at"]
    return servers


@router.get("/{vms_id}")
async def get_vms(vms_id: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"id": vms_id, **resolve_tenant_filter(user, x_tenant_id)}
    server = await db.vms_servers.find_one(query, {"_id": 0})
    if not server:
        raise HTTPException(status_code=404, detail="VMS not found")
    return server


@router.put("/{vms_id}")
async def update_vms(vms_id: str, input: VMSServerUpdate, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    update_data = {k: v for k, v in input.model_dump().items() if v is not None}
    if not update_data:
        raise HTTPException(status_code=400, detail="No data to update")
    query = {"id": vms_id, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.vms_servers.update_one(query, {"$set": update_data})
    if result.matched_count == 0:
        raise HTTPException(status_code=404, detail="VMS not found")
    return await get_vms(vms_id, user, x_tenant_id)


@router.delete("/{vms_id}")
async def delete_vms(vms_id: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"id": vms_id, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.vms_servers.delete_one(query)
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="VMS not found")
    return {"status": "deleted"}


@router.put("/groups/{old_name}")
async def rename_vms_group(old_name: str, req: GroupRenameRequest, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"group_name": old_name, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.vms_servers.update_many(query, {"$set": {"group_name": req.new_name.strip()}})
    return {"updated": result.modified_count}


@router.delete("/groups/{old_name}")
async def delete_vms_group(old_name: str, user: dict = Depends(require_auth), x_tenant_id: Optional[str] = Header(None)):
    query = {"group_name": old_name, **resolve_tenant_filter(user, x_tenant_id)}
    result = await db.vms_servers.update_many(query, {"$unset": {"group_name": ""}})
    return {"updated": result.modified_count}


@router.get("/{vms_id}/test")
async def test_vms_connection(vms_id: str):
    server = await db.vms_servers.find_one({"id": vms_id}, {"_id": 0})
    if not server:
        raise HTTPException(status_code=404, detail="VMS not found")
    
    data = await fetch_vms_data(server, "/rsapi/cameras")
    if data:
        return {"status": "connected", "message": "VMS bağlantısı başarılı"}
    return {"status": "error", "message": "VMS bağlantısı kurulamadı"}


@router.get("/{vms_id}/cameras")
async def fetch_vms_cameras(vms_id: str):
    """Fetch available cameras from VMS server with proper names"""
    from vms_utils import parse_camera_list_xml
    
    server = await db.vms_servers.find_one({"id": vms_id}, {"_id": 0})
    if not server:
        raise HTTPException(status_code=404, detail="VMS not found")
    
    # First, get ALL cameras with names from /rsapi/cameras endpoint
    camera_list_xml = await fetch_vms_data(server, "/rsapi/cameras")
    all_cameras = {}
    
    if camera_list_xml:
        parsed_list = parse_camera_list_xml(camera_list_xml)
        for cam in parsed_list.get('cameras', []):
            all_cameras[cam['camera_id']] = {
                'camera_id': cam['camera_id'],
                'name': cam['name'],
                'description': cam.get('description', ''),
                'disabled': cam.get('disabled', False),
                'model': cam.get('model', ''),
                'has_counter': False,
                'has_queue': False,
                'has_analytics': False,
                'type': 'general'
            }
    
    # Now check which cameras are used in counter module
    counter_data = await fetch_vms_data(server, "/rsapi/modules/counter/getstats")
    counter_camera_ids = set()
    if counter_data:
        parsed = parse_counter_xml(counter_data)
        for cam in parsed.get('cameras', parsed) if isinstance(parsed, dict) else parsed:
            if isinstance(cam, dict):
                cam_id = cam.get('camera_id')
                if cam_id:
                    counter_camera_ids.add(cam_id)
                    if cam_id in all_cameras:
                        all_cameras[cam_id]['has_counter'] = True
                        all_cameras[cam_id]['type'] = 'counter'
                        all_cameras[cam_id]['in_count'] = cam.get('in_count', 0)
                        all_cameras[cam_id]['out_count'] = cam.get('out_count', 0)
                        all_cameras[cam_id]['last_reset'] = cam.get('last_reset', '')
                    else:
                        # Camera in module but not in list - add it
                        all_cameras[cam_id] = {
                            'camera_id': cam_id,
                            'name': f"Sayaç Kamera {cam_id[:8]}",
                            'has_counter': True,
                            'has_queue': False,
                            'has_analytics': False,
                            'type': 'counter',
                            'in_count': cam.get('in_count', 0),
                            'out_count': cam.get('out_count', 0)
                        }
    
    # Check which cameras are used in queue module
    queue_data = await fetch_vms_data(server, "/rsapi/modules/queue/getstats")
    if queue_data:
        parsed = parse_queue_xml(queue_data)
        for cam in parsed.get('cameras', []):
            cam_id = cam.get('camera_id')
            if cam_id:
                if cam_id in all_cameras:
                    all_cameras[cam_id]['has_queue'] = True
                    all_cameras[cam_id]['zones'] = cam.get('zones', [])
                    if all_cameras[cam_id]['type'] == 'general':
                        all_cameras[cam_id]['type'] = 'queue'
                else:
                    all_cameras[cam_id] = {
                        'camera_id': cam_id,
                        'name': f"Kuyruk Kamera {cam_id[:8]}",
                        'has_counter': False,
                        'has_queue': True,
                        'has_analytics': False,
                        'type': 'queue',
                        'zones': cam.get('zones', [])
                    }
    
    # Check which cameras are used in analytics/FR module
    analytics_data = await fetch_vms_data(server, "/rsapi/modules/fr/analytics/getstats")
    if analytics_data:
        parsed = parse_counter_xml(analytics_data)  # Similar XML structure
        for cam in parsed.get('cameras', parsed) if isinstance(parsed, dict) else parsed:
            if isinstance(cam, dict):
                cam_id = cam.get('camera_id')
                if cam_id:
                    if cam_id in all_cameras:
                        all_cameras[cam_id]['has_analytics'] = True
                        if all_cameras[cam_id]['type'] == 'general':
                            all_cameras[cam_id]['type'] = 'analytics'
                    else:
                        all_cameras[cam_id] = {
                            'camera_id': cam_id,
                            'name': f"Analitik Kamera {cam_id[:8]}",
                            'has_counter': False,
                            'has_queue': False,
                            'has_analytics': True,
                            'type': 'analytics'
                        }
    
    # Mark cameras already in DB
    all_vms_ids = list(all_cameras.keys())
    existing_in_db = set()
    if all_vms_ids:
        docs = await db.cameras.find(
            {"camera_vms_id": {"$in": all_vms_ids}}, {"camera_vms_id": 1, "_id": 0}
        ).to_list(1000)
        existing_in_db = {d["camera_vms_id"] for d in docs}

    cameras = list(all_cameras.values())
    for cam in cameras:
        cam["in_db"] = cam["camera_id"] in existing_in_db

    # Sort: enabled first, then by name
    cameras.sort(key=lambda c: (c.get('disabled', False), c.get('name', '')))

    return {
        "vms_id": vms_id,
        "vms_name": server.get("name", ""),
        "cameras": cameras,
        "total": len(cameras)
    }


@router.post("/{vms_id}/import-cameras")
async def import_vms_cameras(vms_id: str, request: ImportCamerasRequest, user: dict = Depends(require_auth)):
    """Import cameras from VMS and save to database"""
    server = await db.vms_servers.find_one({"id": vms_id, **get_tenant_filter(user)}, {"_id": 0})
    if not server:
        raise HTTPException(status_code=404, detail="VMS not found")
    
    # Get cameras list from request
    camera_ids = request.camera_ids
    
    imported = 0
    skipped = 0
    
    # Fetch cameras from VMS first
    cameras_response = await fetch_vms_cameras(vms_id)
    vms_cameras = {c["camera_id"]: c for c in cameras_response.get("cameras", [])}
    
    for cam_id in camera_ids:
        cam_data = vms_cameras.get(cam_id, {"camera_id": cam_id, "type": "counter"})
        
        existing = await db.cameras.find_one({"camera_vms_id": cam_id})
        if existing:
            skipped += 1
            continue
        
        tenant_id = None if user.get("role") == "super_admin" else user.get("tenant_id")
        camera = Camera(
            store_id="",  # Will be assigned later
            camera_vms_id=cam_id,
            name=cam_data.get("name", f"Kamera {cam_id[:8]}"),
            type=cam_data.get("type", "counter"),
            tenant_id=tenant_id,
        )
        doc = camera.model_dump()
        doc['created_at'] = doc['created_at'].isoformat()
        await db.cameras.insert_one(doc)
        imported += 1
    
    return {
        "status": "success",
        "imported": imported,
        "skipped": skipped,
        "message": f"{imported} kamera eklendi, {skipped} kamera zaten mevcut"
    }


@router.post("/{vms_id}/sync-cameras")
async def sync_vms_cameras(vms_id: str, store_id: str = None):
    """Fetch and automatically import all cameras from VMS"""
    server = await db.vms_servers.find_one({"id": vms_id}, {"_id": 0})
    if not server:
        raise HTTPException(status_code=404, detail="VMS not found")
    
    cameras_response = await fetch_vms_cameras(vms_id)
    vms_cameras = cameras_response.get("cameras", [])
    
    if not vms_cameras:
        return {"status": "warning", "message": "VMS'de kamera bulunamadı", "imported": 0}
    
    if store_id:
        stores = [await db.stores.find_one({"id": store_id}, {"_id": 0})]
        stores = [s for s in stores if s]
    else:
        stores = await db.stores.find({"vms_id": vms_id}, {"_id": 0}).to_list(100)
    
    if not stores:
        return {"status": "warning", "message": "Bu VMS'e bağlı lokasyon bulunamadı", "imported": 0}
    
    total_imported = 0
    total_skipped = 0
    
    for store in stores:
        for cam_data in vms_cameras:
            existing = await db.cameras.find_one({
                "camera_vms_id": cam_data["camera_id"],
                "store_id": store["id"]
            })
            
            if existing:
                total_skipped += 1
                continue
            
            cam_type = cam_data.get("type", "counter")
            type_names = {"counter": "Sayaç", "queue": "Kuyruk", "analytics": "Analitik"}
            cam_name = f"{type_names.get(cam_type, 'Kamera')} - {cam_data['camera_id'][:8]}"
            
            camera = Camera(
                store_id=store["id"],
                camera_vms_id=cam_data["camera_id"],
                name=cam_name,
                type=cam_type
            )
            doc = camera.model_dump()
            doc['created_at'] = doc['created_at'].isoformat()
            await db.cameras.insert_one(doc)
            total_imported += 1
    
    return {
        "status": "success",
        "imported": total_imported,
        "skipped": total_skipped,
        "stores_count": len(stores),
        "message": f"{total_imported} kamera eklendi ({len(stores)} mağazaya)"
    }

File: backend-node/routes/vms.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const axios = require('axios');
const { parseStringPromise } = require('xml2js');
const { getDB } = require('../db');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();

async function fetchVmsData(server, endpoint) {
  try {
    const url = `${server.url.replace(/\/$/, '')}${endpoint}`;
    const response = await axios.get(url, {
      auth: { username: server.username, password: server.password || '' },
      timeout: 10000
    });
    return response.data;
  } catch {
    return null;
  }
}

router.post('/', requireAuth, async (req, res) => {
  const db = getDB();
  const doc = { id: uuidv4(), is_active: true, created_at: new Date().toISOString(), password: '', ...req.body };
  await db.collection('vms_servers').insertOne(doc);
  const { _id, ...out } = doc;
  res.json(out);
});

router.get('/', requireAuth, async (req, res) => {
  const db = getDB();
  const servers = await db.collection('vms_servers').find({}, { projection: { _id: 0 } }).toArray();
  res.json(servers);
});

router.get('/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const server = await db.collection('vms_servers').findOne({ id: req.params.id }, { projection: { _id: 0 } });
  if (!server) return res.status(404).json({ detail: 'VMS bulunamadı' });
  res.json(server);
});

router.put('/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const update = Object.fromEntries(Object.entries(req.body).filter(([, v]) => v !== undefined && v !== null));
  const result = await db.collection('vms_servers').updateOne({ id: req.params.id }, { $set: update });
  if (result.matchedCount === 0) return res.status(404).json({ detail: 'VMS bulunamadı' });
  const updated = await db.collection('vms_servers').findOne({ id: req.params.id }, { projection: { _id: 0 } });
  res.json(updated);
});

router.delete('/:id', requireAuth, async (req, res) => {
  const db = getDB();
  const result = await db.collection('vms_servers').deleteOne({ id: req.params.id });
  if (result.deletedCount === 0) return res.status(404).json({ detail: 'VMS bulunamadı' });
  res.json({ status: 'deleted' });
});

router.get('/:id/test', requireAuth, async (req, res) => {
  const db = getDB();
  const server = await db.collection('vms_servers').findOne({ id: req.params.id }, { projection: { _id: 0 } });
  if (!server) return res.status(404).json({ detail: 'VMS bulunamadı' });
  const data = await fetchVmsData(server, '/rsapi/cameras');
  if (data) return res.json({ status: 'connected', message: 'VMS bağlantısı başarılı' });
  res.json({ status: 'error', message: 'VMS bağlantısı kurulamadı' });
});

async function parseCameraListXml(xml) {
  try {
    const allCameras = {};
    const parsed = await parseStringPromise(xml, { explicitArray: false, ignoreAttrs: false });

    // Olası kök elementleri dene
    const root = parsed?.Cameras || parsed?.cameras || parsed?.CameraList || parsed?.Items || parsed;
    let list = root?.Camera || root?.camera || root?.Item || root?.item || [];
    if (!Array.isArray(list)) list = list ? [list] : [];

    for (const cam of list) {
      const attrs = cam?.$ || {};
      const id = cam?.ID || cam?.Id || cam?.id || cam?.CameraID || cam?.cameraId || attrs?.id || attrs?.ID;
      if (!id) continue;
      const name = cam?.Name || cam?.name || attrs?.name || attrs?.Name || `Kamera ${String(id).substring(0, 8)}`;
      const disabled = (cam?.Disabled || cam?.disabled || 'false').toString().toLowerCase() === 'true';
      allCameras[String(id).trim()] = {
        camera_id: String(id).trim(), name: String(name).trim(),
        disabled, has_counter: false, has_queue: false, has_analytics: false, type: 'general'
      };
    }
    return allCameras;
  } catch { return {}; }
}

function parseCounterXml(xml) {
  const cameras = {};
  const stateRegex = /<CameraState>([\s\S]*?)<\/CameraState>/g;
  let m;
  while ((m = stateRegex.exec(xml)) !== null) {
    const b = m[1];
    const id = b.match(/<CameraID>([^<]+)<\/CameraID>/)?.[1]?.trim();
    if (!id) continue;
    const inCount = parseInt(b.match(/<In>([^<]+)<\/In>/)?.[1] || '0');
    const outCount = parseInt(b.match(/<Out>([^<]+)<\/Out>/)?.[1] || '0');
    cameras[id] = { in_count: inCount, out_count: outCount };
  }
  return cameras;
}

function parseQueueXml(xml) {
  const cameras = {};
  const stateRegex = /<CameraState>([\s\S]*?)<\/CameraState>/g;
  let m;
  while ((m = stateRegex.exec(xml)) !== null) {
    const b = m[1];
    const id = b.match(/<CameraID>([^<]+)<\/CameraID>/)?.[1]?.trim();
    if (!id) continue;
    cameras[id] = { zones: [] };
  }
  return cameras;
}

router.get('/:id/cameras', requireAuth, async (req, res) => {
  const db = getDB();
  const server = await db.collection('vms_servers').findOne({ id: req.params.id }, { projection: { _id: 0 } });
  if (!server) return res.status(404).json({ detail: 'VMS bulunamadı' });

  const allCameras = {};

  // Kamera listesini çek
  const cameraListXml = await fetchVmsData(server, '/rsapi/cameras');
  if (cameraListXml) {
    const parsed = await parseCameraListXml(cameraListXml);
    Object.assign(allCameras, parsed);
  }

  // Sayaç modülü kameralarını işaretle
  const counterXml = await fetchVmsData(server, '/rsapi/modules/counter/getstats');
  if (counterXml) {
    const counterCams = parseCounterXml(counterXml);
    for (const [id, data] of Object.entries(counterCams)) {
      if (allCameras[id]) {
        allCameras[id].has_counter = true;
        allCameras[id].type = 'counter';
        allCameras[id] = { ...allCameras[id], ...data };
      } else {
        allCameras[id] = { camera_id: id, name: `Sayaç Kamera ${id.substring(0, 8)}`, has_counter: true, has_queue: false, has_analytics: false, type: 'counter', ...data };
      }
    }
  }

  // Kuyruk modülü kameralarını işaretle
  const queueXml = await fetchVmsData(server, '/rsapi/modules/queue/getstats');
  if (queueXml) {
    const queueCams = parseQueueXml(queueXml);
    for (const [id] of Object.entries(queueCams)) {
      if (allCameras[id]) {
        allCameras[id].has_queue = true;
        if (allCameras[id].type === 'general') allCameras[id].type = 'queue';
      } else {
        allCameras[id] = { camera_id: id, name: `Kuyruk Kamera ${id.substring(0, 8)}`, has_counter: false, has_queue: true, has_analytics: false, type: 'queue' };
      }
    }
  }

  // Yaş/Cinsiyet (FR/Analytics) modülü kameralarını işaretle
  const frXml = await fetchVmsData(server, '/rsapi/modules/fr/analytics/getstats');
  if (frXml) {
    const frCams = parseCounterXml(frXml);
    for (const [id] of Object.entries(frCams)) {
      if (allCameras[id]) {
        allCameras[id].has_analytics = true;
        if (allCameras[id].type === 'general') allCameras[id].type = 'analytics';
      } else {
        allCameras[id] = { camera_id: id, name: `Analitik Kamera ${id.substring(0, 8)}`, has_counter: false, has_queue: false, has_analytics: true, type: 'analytics' };
      }
    }
  }

  const cameras = Object.values(allCameras).sort((a, b) => (a.disabled ? 1 : 0) - (b.disabled ? 1 : 0) || a.name.localeCompare(b.name));
  res.json({ vms_id: req.params.id, vms_name: server.name, cameras, total: cameras.length });
});

router.post('/:id/import-cameras', requireAuth, async (req, res) => {
  const db = getDB();
  const server = await db.collection('vms_servers').findOne({ id: req.params.id }, { projection: { _id: 0 } });
  if (!server) return res.status(404).json({ detail: 'VMS bulunamadı' });

  const { camera_ids = [], store_id = '' } = req.body;
  let imported = 0, skipped = 0;

  // VMS'den güncel kamera bilgilerini çek (isim ve tip için)
  const allCameras = {};
  const cameraListXml = await fetchVmsData(server, '/rsapi/cameras');
  if (cameraListXml) Object.assign(allCameras, await parseCameraListXml(cameraListXml));
  const counterXml = await fetchVmsData(server, '/rsapi/modules/counter/getstats');
  if (counterXml) { for (const [id] of Object.entries(parseCounterXml(counterXml))) { if (allCameras[id]) { allCameras[id].has_counter = true; allCameras[id].type = 'counter'; } } }
  const queueXml = await fetchVmsData(server, '/rsapi/modules/queue/getstats');
  if (queueXml) { for (const [id] of Object.entries(parseQueueXml(queueXml))) { if (allCameras[id]) { allCameras[id].has_queue = true; if (allCameras[id].type === 'general') allCameras[id].type = 'queue'; } } }
  const frXml = await fetchVmsData(server, '/rsapi/modules/fr/analytics/getstats');
  if (frXml) { for (const [id] of Object.entries(parseCounterXml(frXml))) { if (allCameras[id]) { allCameras[id].has_analytics = true; if (allCameras[id].type === 'general') allCameras[id].type = 'analytics'; } } }

  for (const camId of camera_ids) {
    const existing = await db.collection('cameras').findOne({ camera_vms_id: camId, store_id });
    if (existing) { skipped++; continue; }
    const camInfo = allCameras[camId] || {};
    await db.collection('cameras').insertOne({
      id: uuidv4(),
      store_id,
      camera_vms_id: camId,
      name: camInfo.name || `Kamera ${camId.substring(0, 8)}`,
      type: camInfo.type || 'counter',
      is_active: true,
      created_at: new Date().toISOString()
    });
    imported++;
  }

  res.json({ status: 'success', imported, skipped, message: `${imported} kamera eklendi, ${skipped} kamera zaten mevcut` });
});

module.exports = router;
